test/EndToEndTests.cpp

```diff
@@ -0,0 +1,47 @@
+#include "../OP2Utility/include/OP2Utility.h"
+#include <gtest/gtest.h>
+#include <string>
+
+using namespace OP2Utility;
+
+std::string getMsvcExeDirectory();
+
+TEST(EndToEnd, ImageConversion)
+{
+
+#ifdef _WIN32
+	const std::string applicationName = getMsvcExeDirectory() + "OP2BitmapConverter.exe";
+#else // Linux makefile
+	const std::string applicationName = "";
+#endif
+
+	const std::string startingTilesetFilename("TilesetTest.bmp");
+	const std::string customTilesetFilename("CustomTilesetTest.bmp");
+	const std::string standardTilesetFilename("StandardTilesetTest.bmp");
+
+	// Convert standard 8-bit bitmap into custom tileset format
+	std::system((applicationName + " " + startingTilesetFilename + " " + customTilesetFilename).c_str());
+	
+	// Convert custom tileset format back into custom 8-bit bitmap
+	std::system((applicationName + " " + customTilesetFilename + " " + standardTilesetFilename).c_str());
+
+
+	auto bitmap1 = Tileset::ReadTileset(Stream::FileReader(startingTilesetFilename));
+	auto bitmap2 = Tileset::ReadTileset(Stream::FileReader(standardTilesetFilename));
+
+	ASSERT_EQ(bitmap1, bitmap2);
+}
+
+// Return the relative directory where OP2BitmapConverter is compiled
+std::string getMsvcExeDirectory()
+{
+#if defined(_WIN64) && defined(_DEBUG)
+	return "..\\x64\\debug\\";
+#elif _WIN64
+	return "..\\x64\\release\\";
+#elif _DEBUG
+	return "..\\debug\\";
+#else
+	return "..\\release\\";
+#endif
+}
```

Comment:
I'm a little concerned by this part. I feel like this level of detail may be more appropriate as part of project settings, rather than source code within the test project.

The target executable paths depend on project configuration, but the source code itself doesn't have access to that information. The actual target executable output paths can be modified in the project settings, or new configurations added, and there is no way for the test code to really know that. It would be unfortunate to require source code updates that corresponded with project configuration updates.

Not only that, but the method name is very specific, with `Msvc` in it, and a similar guard around the call point. At the very least, it would be nice to have one method name to get the exe directory, and put all platform detects in that one method.

----

In terms of moving this out of the source code, and into project config, I image a test harness that might do something like:
```bash
export PATH="${PATH}:${MainExePath}"
cd "${TestDataPath}"
${TestExe}
```

Another option might be to have the test executable take parameters:
```bash
${TestExe} "${MainExe}" "${TestDataPath}"
```

----

It seems these end-to-end tests are somewhat dependent on environment (finding executable to test, and data to test it with). That means it's a little more involved that just double clicking a compiled executable. We need some kind of environment setup, either through environment variables, or through parameters passed to the exe. To keep things simple with one file to run, we'd need to package that environment into a run script to setup that environment and run the test executable, or somehow place that environment information into project settings so it gets passed to the test executable during startup.

Reply:
I can setup environment variables during the post build event on compilation for MSVC builds which would have access to the project path macros. I wasn't sure I was comfortable changing environment variables on someone's machine if they download the project and compile. If you think this is ok, I'll go ahead and try out a post build script to setup a couple of environment variables. 

No problem further isolating the code to determine the executable path. I agree the current solution is brittle, but it will fail in obvious ways immediately after changing the project settings and it is a trivial fix in one location to correct.

Reply:
Environment variables are limited to shell sessions (and inherited by subsessions). You can't affect the environment variables of a parent shell. Such environment variables will have limited effect in the shell spawned to run the test executable. It will not have system wide effect.

Reply:
I spent some time this morning prototyping. It looks like variables set using the console `SET` command does not persist from building the application to building the unit test code. I am pretty rusty at command line so possibility of operator error.

Post build event for application build process:
```
set OP2TilesetConverterPath="$(TargetPath)"
```

Then in the unit test, I tried calling `char* std::getenv( const char* env_var )`. This returned a nullptr, indicating the environment variable was not set. I noticed there is a Microsoft specific function to get environment variables via a win32 call, but I didn't try it.

As an alternative, I am saving the application's path to a .txt file during the post build event and placing it in the test code directory for the test code to scrape. This is working. I'll upload shortly to verify that appveyor handles it ok in automated build settings.

Reply:
This is likely a consequence of environment variables only persisting until the shell is terminated. Each command invoked by Visual Studio may start a new fresh shell. It's not surprising that any environment variables set while building the application are no longer present when running the test code.

Does the test project produce an executable in the same folder as the main executable? Maybe the test project variables are sufficient to find the main executable. Might be easier to use those variables in a context where they would be remembered when running the test executable.

Reply:
Hmm, just came across this, which seems relevant:
[How to get the TargetPath from ProjectReference](https://social.msdn.microsoft.com/Forums/vstudio/en-US/ff6233b2-92b0-482d-afb7-14aaea82f10a/how-to-get-the-targetpath-from-projectreference?forum=msbuild)